relations/src/shielder/tangle.rs

```diff
@@ -1,67 +1,81 @@
 //! All the index intervals used here are closed-open, i.e. they are in form `[a, b)`, which means
 //! that we consider indices `a`, `a+1`, ..., `b-1`. We also use 0-based indexing.
 
-use ark_r1cs_std::R1CSVar;
+use std::ops::Add;
+
+use ark_ff::{Field, Zero};
+use ark_r1cs_std::fields::FieldVar;
 use ark_relations::r1cs::SynthesisError;
-use ark_std::vec::Vec;
 
-use super::types::ByteVar;
+use crate::{environment::FpVar, CircuitField};
 
 /// Bottom-level chunk length.
 const BASE_LENGTH: usize = 4;
+/// Tangling operates on fixed-sized arrays with 128 elements.
+const EXPAND_TO: usize = 128;
 
-/// Tangle elements of `bytes`.
+/// Entangle `input` into single `FpVar`.
 ///
 /// For circuit use only.
-pub(super) fn tangle_in_field<const SQUASH_FACTOR: usize>(
-    mut bytes: Vec<ByteVar>,
-) -> Result<Vec<ByteVar>, SynthesisError> {
-    let number_of_bytes = bytes.len();
-    _tangle_in_field(&mut bytes, 0, number_of_bytes)?;
-    Ok(bytes
-        .chunks(SQUASH_FACTOR)
-        .map(|chunk| {
-            chunk
-                .iter()
-                .cloned()
-                .reduce(|x, y| x.xor(&y).unwrap())
-                .unwrap()
-        })
-        .collect())
+pub(super) fn tangle_in_circuit(input: &[FpVar]) -> Result<FpVar, SynthesisError> {
+    let mut input_expanded = input
+        .iter()
+        .cycle()
+        .take(EXPAND_TO)
+        .cloned()
+        .collect::<Vec<_>>();
+
+    _tangle_in_circuit(&mut input_expanded, 0, EXPAND_TO)?;
+
+    Ok(input_expanded.into_iter().reduce(|a, b| a.add(b)).unwrap())
+}
+
+fn dezeroize_in_circuit(
+    fp: &FpVar,
+    fallback: impl Into<CircuitField>,
+) -> Result<FpVar, SynthesisError> {
+    let fallback = FpVar::constant(fallback.into());
+    fp.is_zero()?.select(&fallback, fp)
 }
 
 /// Recursive and index-bounded implementation of the first step of the `tangle` procedure.
-fn _tangle_in_field(bytes: &mut [ByteVar], low: usize, high: usize) -> Result<(), SynthesisError> {
-    // Bottom level case: computing suffix sums. We have to do some loop-index boilerplate, because
-    // Rust doesn't support decreasing range iteration.
+fn _tangle_in_circuit(elems: &mut [FpVar], low: usize, high: usize) -> Result<(), SynthesisError> {
+    // Bottom level case: computing suffix sums of inverses. We have to do some loop-index
+    // boilerplate, because Rust doesn't support decreasing range iteration.
     if high - low <= BASE_LENGTH {
         let mut i = high - 2;
         loop {
-            bytes[i] = ByteVar::constant(
-                u8::overflowing_add(
-                    bytes[i].value().unwrap_or_default(),
-                    bytes[i + 1].value().unwrap_or_default(),
-                )
-                .0,
-            );
+            let previous = dezeroize_in_circuit(&elems[i + 1], (2 * low + 1) as u64)?;
+            let current = dezeroize_in_circuit(&elems[i], (2 * high + 1) as u64)?;
+
+            elems[i] = (previous.inverse()? + current.inverse()?)
```

Comment:
Use `1/a + 1/b = (a+b)/ab` as this reduces 2 costly inverses to 1.  Also, if you `dezeroize_in_circuit` here if `a+b` then you don't need to `dezeroize_in_circuit` later as this is a prime field ;)

Reply:
fair point, but I chose to remove inverting `previous` - now it is just suffix sums of inverses (with some const scaling); when I started reworking towards `(a+b)/ab` the code looked too much complicated for its quality xD